BrainPortal/lib/boutiques_dir_maker.rb

```diff
@@ -0,0 +1,166 @@
+    descriptor = super.dup
+    boutiques_dir_maker_list = descriptor.custom_module_info('BoutiquesDirMaker')
+
+    boutiques_dir_maker_list.each do |inputid,dirname|
+      # Adjust the description
+      input        = descriptor.input_by_id(inputid)
+      description  = input.description.presence || ""
+      description += "\n" if description
+      description += "This input will be copied in a parent folder `#{dirname}`.\n The parent folder will be used in the command line"
+      input.description = description
+    end
+
+    descriptor
+  end
+
+  ############################################
+  # Bourreau (Cluster) Side Modifications
+  ############################################
+
+  # For input in the `BoutiquesDirMaker` section,
+  # create a directory based on the value BoutiquesDirMaker[input_id]
+  # that will contains a symlink to the orginal selected Userfile.
+  def setup #:nodoc:
+    return false unless super
+
+    descriptor = self.descriptor_for_setup
+    self.addlog(descriptor.file_revision_info.format("%f rev. %s %a %d"))
+
+    boutiques_dir_maker_list = descriptor.custom_module_info('BoutiquesDirMaker')
+    descriptor.file_inputs.each do |input|
+      userfile_id = invoke_params[input.id]
+      next if userfile_id.blank? # that happens when it's an optional file
+      userfile    = Userfile.find(userfile_id)
+      dirname     = boutiques_dir_maker_list[input.id]
+
+      # Most common situation
+      if ! input.list || ! userfile.is_a?(CbrainFileList)
+        make_available(userfile, dirname.present? ? "#{dirname}/#{userfile.name}" : userfile.name)
+        next
+      end
+
+      # In case the input is a list and is assigned a CbrainFileList
+      userfile.sync_to_cache
+      userfile_list = userfile.userfiles_accessible_by_user!(user, nil, nil, file_access_symbol)
+      userfile_list.compact.each do |subfile|
+        make_available(subfile, subfile.name)
+      end
+    end
+
+    true
+  end
+
+  # It adjusts the command-line of the descriptor.
+  # So that the token for the parent directory is replaced by the
+  # name of the parent directory we created in setup().
+  #
+  # From:
+  #
+  #   "command-line": "apptool [PRECOMPUTED_INPUT] [OUTPUT] stuff"
+  #
+  # To:
+  #
+  #   "command-line": "true [PRECOMPUTED_INPUT]; apptool #{parent_directory}/#{usefile.name} [OUTPUT] stuff"
+  #
+  # The reason a dummy true statement is prefixed at the beginning of the command
+  # is so that bosh won't complain if it can't find the token [PRECOMPUTED_INPUT] anywhere
+  # in the string.
+  def descriptor_for_cluster_commands
+    descriptor = super.dup
+
+    command    = descriptor.command_line
+
+    boutiques_dir_maker_list = descriptor.custom_module_info('BoutiquesDirMaker')
+    boutiques_dir_maker_list.each do |inputid,dirname|
+      input         = descriptor.input_by_id(inputid)
+      token         = input.value_key # e.g. '[PRECOMPUTED_INPUT]'
+      userfile_id   = invoke_params[inputid]
```

Comment:
This is all wrong, you're substituting filenames inside the command line before bosh is given the task to do it. Bosh will complain and maybe crash when it doesn't find the substitution tokens it wants. If you're modifying the command-line, why not just replace

```ruby
   "blah blah [PRECOMPUTED_INPUT] blah blah"
```

by

```ruby
   "blah blah #{dirname.bash_escape}/[PRECOMPUTED_INPUT] blah blah"
```

that way the the substitution is still the responibility of bosh?

Also, the way you implemented it, doing a string substitution inside the command line, is a security risk for code injection! You can't just substitute things under user control without escaping.